### PR TITLE
Fix cargo-bazel recompile for MODULE.bazel

### PR DESCRIPTION
**Pull request by ericmcbride**
- Addresses https://github.com/bazelbuild/rules_rust/issues/2531
- Wrote a function, thats basically copy / pasted from `crates_universe/private/generate_utils.bzl` to account for the `module_ctx`  Can probably make the original function more generic if thats more desired.  I'm not sure if the maintainers would prefer a one off, since the `MODULE.bazel `feature set for `crates_universe` isn't exactly "Stable" yet.  
- Noticed theres no way to "bootstrap" the new `cargo` features in `MODULE.bazel` , like you can for `crates_universe`.   May need to open an issue for this.  Tested by using the `CARGO_BAZEL_GENERATOR_URL` and `CARGO_BAZEL_GENERATOR_SHA256` environmentals.  This looks to cause some failing unit tests.  Not sure how to get support on this
- Failsafe of using the cargo buildstrap `cargo-bazel` if no URLs are found, and if theres no environmentals.

### CODE REVIEW

File: crate_universe/private/module_extensions/cargo_bazel_bootstrap.bzl [hunk 1]
```diff
@@ -15,12 +15,12 @@ cargo_bazel_bootstrap = module_extension(
...
 )
 
-def get_cargo_bazel_runner(module_ctx):
+def get_cargo_bazel_runner(module_ctx, cargo_bazel):
     """A helper function to allow executing cargo_bazel in module extensions.
 
     Args:
         module_ctx: The module extension's context.
-
+        cargo_bazel: (tuple, dict) The path to a `cargo-bazel` binary
```
Comment (dzbarsky) [inline crate_universe/private/module_extensions/cargo_bazel_bootstrap.bzl:23]: nit: the type should be `Path`, right?

File: crate_universe/extension.bzl [hunk 2]
```diff
@@ -181,8 +183,60 @@ def _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg, annotations):
...
+        )
+        return output
+
+    result = module_ctx.download(
+        output = output,
+        url = generator_url,
+        executable = True,
+    )
+
+    return output
```
Comment (UebelAndre) [inline crate_universe/extension.bzl:234]: You could build kwargs here
```suggestion
    download_kwargs = {
        "output": output,
        "url": generator_url,
        "executable": True,
    }
    
    if generator_sha256:
        download_kwargs.update({"sha256": generator_sha256})

    module_ctx.download(**download_kwargs)
    return output
```
I think this format also reveals that result isn't being used. Should it be used?

File: crate_universe/extension.bzl [hunk 3]
```diff
@@ -181,8 +183,60 @@ def _generate_hub_and_spokes(module_ctx, cargo_bazel, cfg, annotations):
         else:
             fail("Invalid repo: expected Http or Git to exist for crate %s-%s, got %s" % (name, version, repo))
 
+def _get_generator(module_ctx):
+    """ Query Network Resources to local a `cargo-bazel` binary.  Based off get_generator in
```
Comment (UebelAndre) [inline crate_universe/extension.bzl:187]:
```suggestion
    """Query Network Resources to local a `cargo-bazel` binary.  Based off get_generator in 
```

**Comment by UebelAndre** [comment]
> * Noticed theres no way to "bootstrap" the new `cargo` features in `MODULE.bazel` , like you can for `crates_universe`.   May need to open an issue for this.  Tested by using the `CARGO_BAZEL_GENERATOR_URL` and `CARGO_BAZEL_GENERATOR_SHA256` environmentals.  This looks to cause some failing unit tests.  Not sure how to get support on this

This can probably be spawned into a separate issue.

File: crate_universe/extension.bzl (hunk 2 above)
Comment (ericmcbride) [reply crate_universe/extension.bzl:234]: Ah yeah i see that now.  I can make these adjustments and i def prefer the above.  What do I need to do, to be able to generate the URLS for the tests?  Ive been scouring and cannot figure out how to do this.  I know the file is generated and bundled with releases.  I was able to test with the environmental, but I dont think theres a current way to bootstrap the `MODULE.bazel` crates stuff yet.

Comment (UebelAndre) [reply crate_universe/extension.bzl:234]: I don't think we currently have a test for this. If you wanted to add one it'd likely be creating a new job that's using `shell_commands` to build a `cargo_bazel` binary and then generate a `.bazelrc` file which uses `--repo_env` to set the two variables and see if you can build a minimal repo.

Comment (ericmcbride) [reply crate_universe/extension.bzl:234]: Cool thanks!

**Review by UebelAndre — APPROVED** [review]
Thanks!

**Comment by UebelAndre** [comment]
> > * Noticed theres no way to "bootstrap" the new `cargo` features in `MODULE.bazel` , like you can for `crates_universe`.   May need to open an issue for this.  Tested by using the `CARGO_BAZEL_GENERATOR_URL` and `CARGO_BAZEL_GENERATOR_SHA256` environmentals.  This looks to cause some failing unit tests.  Not sure how to get support on this
> 
> This can probably be spawned into a separate issue.

Would you also be willing to write up a new issue for this?

**Comment by UebelAndre** [comment]
I also might have messed up resolving the merge conflict. Would you be able to take a look?

**Comment by ericmcbride** [comment]
> I also might have messed up resolving the merge conflict. Would you be able to take a look?

Yes.  I also see master is failing, maybe from a different reason?

**Pull request merged**